hawc/apps/common/models.py

```diff
@@ -340,6 +340,119 @@ def get_maximum_depth(cls, assessment_id):
+        """
+
+        def prettify_indices(list_of_indices: List[int]):
+            """
+            Given a list containing just numeric indices, returns a dot-delimited version with the
+            indices wrapped in square brackets.
+
+            For example,
+            [8, 6, 7] -> "[8].[6].[7]"
+
+            Args:
+                list_of_indices (List[int]): list of indices
+
+            Returns:
+                str: formatted string
+            """
+            return ".".join([f"[{x}]" for x in list_of_indices])
+
+        valid_keys = ["name", "slug", "children"]
+        idx = 0
+        for tag in tags_at_level:
+            tag_name = None
+            slug = None
+            if "name" in tag:
+                tag_name = tag["name"]
+                if "slug" in tag:
+                    # validate it
+                    slug = tag["slug"]
+                    recomputed_slug = default_slugify(slug)
+                    if recomputed_slug != slug:
+                        adjusted_crumbs = prettify_indices(breadcrumbs + [idx])
+                        issues.append(
+                            f"tag '{tag_name}' at index {adjusted_crumbs} has an invalid slug '{slug}'; should be something like '{default_slugify(tag_name)}' or '{recomputed_slug}'"
+                        )
+                else:
+                    # generate it
+                    slug = default_slugify(tag_name)
+                    tag["slug"] = slug
+            else:
+                adjusted_crumbs = prettify_indices(breadcrumbs + [idx])
+                issues.append(f"tag at index {adjusted_crumbs} is missing a name")
+
+            # recurse
+            if "children" in tag:
+                cls.validate_and_transform_tag_tree(tag["children"], issues, breadcrumbs + [idx])
+
+            # delete any extraneous keys that were passed in (could instead/also flag it as an issue?).
+            # why not flag an issue? Sometimes it's nice in JSON to change a key to "_oldname" or something,
+            # since JSON doesn't support comments. This way we don't break if the caller does something like that.
+            keys_to_delete = []
+            for key in tag:
+                if key not in valid_keys:
+                    # need to build up a list; we can't delete from the tag while we iterate through it
+                    keys_to_delete.append(key)
+
+            for key in keys_to_delete:
+                del tag[key]
+
+            # and now restructure the tag so that name/slug are keys inside a "data" object as Treebeard requires.
+            if "name" in tag:
+                del tag["name"]
+            if "slug" in tag:
+                del tag["slug"]
+            tag["data"] = {"name": tag_name, "slug": slug}
+
+            idx += 1
+
+    @classmethod
+    @transaction.atomic
+    def replace_tag_tree_in_assessment(cls, assessment, tag_tree: List[object]):
```

Comment:
use `Dict` or `Any` instead of object; use `assessment_id: int` as well; rename to `replace_tree`